worker/pubsub/remoteserver.go

```diff
@@ -0,0 +1,336 @@
+// RemoteServer represents the public interace of the worker
+// responsible for forwarding messages to a single other API server.
+type RemoteServer interface {
+	worker.Worker
+	UpdateAddresses(addresses []string)
+	Publish(message *params.PubSubMessage)
+}
+
+// remoteServer is responsible for taking messages and sending them to the
+// pubsub endpoint on the remote server. If the connection is dropped, the
+// remote server will try to reconnect. Messages are not sent until the
+// connection either succeeds the first time, or fails to connect. Once there
+// is a failure, incoming messages are dropped until reconnection is complete,
+// then messages will flow again.
+type remoteServer struct {
+	origin string
+	target string
+	info   *api.Info
+
+	openAPI    func(*api.Info, api.DialOpts) (api.Connection, error)
+	openWriter func(api.Connection) (pubsubapi.MessageWriter, error)
+	connection *remoteConnection
+
+	hub   *pubsub.StructuredHub
+	tomb  tomb.Tomb
+	clock clock.Clock
+	mutex sync.Mutex
+
+	pending        *deque.Deque
+	closed         chan struct{}
+	data           chan struct{}
+	stopConnecting chan struct{}
+
+	unsubscribe func()
+}
+
+// RemoteServerConfig defines all the attributes that are needed for a RemoteServer.
+type RemoteServerConfig struct {
+	// Hub is used to publish connection messages
+	Hub    *pubsub.StructuredHub
+	Origin string
+	Target string
+	Clock  clock.Clock
+
+	// APIInfo is initially populated with the addresses of the target machine.
+	APIInfo    *api.Info
+	APIOpen    func(*api.Info, api.DialOpts) (api.Connection, error)
+	WriterOpen func(api.Connection) (pubsubapi.MessageWriter, error)
+}
+
+// NewRemoteServer creates a new RemoteServer that will connect to the remote
+// apiserver and pass on messages to the pubsub endpoint of that apiserver.
+func NewRemoteServer(config RemoteServerConfig) (RemoteServer, error) {
+	closed := make(chan struct{})
+	close(closed)
+	remote := &remoteServer{
+		origin:     config.Origin,
+		target:     config.Target,
+		info:       config.APIInfo,
+		openAPI:    config.APIOpen,
+		openWriter: config.WriterOpen,
+		hub:        config.Hub,
+		clock:      config.Clock,
+		pending:    deque.New(),
+		closed:     closed,
+		data:       make(chan struct{}),
+	}
+	unsub, err := remote.hub.Subscribe(forwarder.ConnectedTopic, remote.onForwarderConnection)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	remote.unsubscribe = unsub
+	go func() {
+		defer remote.tomb.Done()
+		remote.tomb.Kill(remote.loop())
+	}()
+	return remote, nil
+}
+
+func (r *remoteServer) onForwarderConnection(topic string, details forwarder.OriginTarget, err error) {
+	if err != nil {
+		// This should never happen.
+		logger.Errorf("subscriber callback error: %v", err)
+		return
+	}
+	if details.Target == r.origin && details.Origin == r.target {
+		// If we have just been connected to by the apiserver that we are
+		// trying to connect to, interrupt any waiting we may be doing and try
+		// again.
```

Comment:
Our reconnect loop may be in a five minute wait before trying again. If our target connected to us, it means they are now alive. So stop waiting and try again immediately.